File: apps/workspace-engine/pkg/model/repository.go
package model

import (
	"context"
)

type Repository[T Entity] interface {
	GetAll(ctx context.Context) []*T
	Get(ctx context.Context, entityID string) *T
	Create(ctx context.Context, entity *T) error
	Update(ctx context.Context, entity *T) error
	Delete(ctx context.Context, entityID string) error
	Exists(ctx context.Context, entityID string) bool
}

var _ Repository[Entity] = (*RepositoryWithID[Entity])(nil)

type RepositoryWithID[T Entity] struct {
	entities map[string]*T
}

func (r *RepositoryWithID[T]) GetAll(ctx context.Context) []*T {
	entities := make([]*T, 0, len(r.entities))
	for _, entity := range r.entities {
		entities = append(entities, entity)
	}
	return entities
}

func (r *RepositoryWithID[T]) Get(ctx context.Context, entityID string) *T {
	return r.entities[entityID]
}

func (r *RepositoryWithID[T]) Create(ctx context.Context, entity *T) error {
	r.entities[(*entity).GetID()] = entity
	return nil
}

func (r *RepositoryWithID[T]) Update(ctx context.Context, entity *T) error {
	r.entities[(*entity).GetID()] = entity
	return nil
}

func (r *RepositoryWithID[T]) Delete(ctx context.Context, entityID string) error {
	delete(r.entities, entityID)
	return nil
}

func (r *RepositoryWithID[T]) Exists(ctx context.Context, entityID string) bool {
	_, exists := r.entities[entityID]
	return exists
}

File: apps/event-queue/src/repository.ts
type Entity = { id: string };

interface Repository<T extends Entity> {
  get(id: string): T | null;
  getAll(): T[];
  create(entity: T): T;
  update(entity: T): T;
  delete(id: string): T | null;
  exists(id: string): boolean;
}

export class RepositoryWithID<T extends Entity> implements Repository<T> {
  private entities: Record<string, T> = {};

  get(id: string): T | null {
    return this.entities[id] ?? null;
  }
  getAll(): T[] {
    return Object.values(this.entities);
  }
  create(entity: T): T {
    this.entities[entity.id] = entity;
    return entity;
  }
  update(entity: T): T {
    const existing = this.entities[entity.id] ?? {};
    this.entities[entity.id] = { ...existing, ...entity };
    return this.entities[entity.id]!;
  }
  delete(id: string): T | null {
    const entity = this.entities[id];
    if (entity == null) return null;
    delete this.entities[id];
    return entity;
  }
  exists(id: string): boolean {
    return this.entities[id] != null;
  }
}
